### ISSUE TITLE
Multiline Expressions in Indented Syntax

### ISSUE THREAD

**Issue by talsafran** [sass/sass#216]
_Edited by @nex3 to change the title._

[cross-posted from http://stackoverflow.com/questions/8248976/sass-indented-syntax-on-multiple-lines]

I love indented syntax, I think it's much cleaner.

There's one issue I have with it. If I have a really long mixin declaration, there's no way to split it into multiple lines (obeying the 80 character limit, for example)

Take this example, first written in SCSS.

``` scss
@mixin col($cols, $mleft: 0, $mright: 0, $include-margin: false, $border: 0,
           $pleft: 0, $pright: 0, $include-padding: true, $extra: 0, 
           $clear: false, $lead: true, $container: false) {
    color: red;
    display: block;
}
```

I'm able to split up one long declaration in to multiple lines. With the indented syntax, I don't think there's a way. I have to put the declaration on one line, which is way less readable.

``` sass
@mixin col($cols, $mleft: 0, $mright: 0, $include-margin: false, $border: 0, $pleft: 0, $pright: 0, $include-padding: true, $extra: 0, $clear: false, $lead: true, $container: false)
    color: red
    display: block
```

Is there some way I don't know of? And if not, can there be one? :)

**Comment by miracle2k** [comment]
I also think sass should support this, and not just in mixins. It's a serious deficiency in the indented syntax.

**Comment by pengwynn** [comment]
+1 Haml allows multi-line statements if lines end with a comma.

**Comment by talsafran** [comment]
@pengwynn I like that approach. It's nicer than the way I was thinking, which was adding a \ to the end of line like in Python.

**Comment by cybersquid** [comment]
Just ran into this myself.  It's a real problem.  box-shadow statements can be really long.

**Comment by Anahkiasen** [comment]
+1

**Comment by while0pass** [comment]
+1
another example of currently valid, but not very readable src property in @font-face rule

```
@font-face
    font-family: 'SPEdessa'
    src: url('fonts/spedessa-webfont.eot')
    src: url('fonts/spedessa-webfont.eot?#iefix') format('embedded-opentype'), url('fonts/spedessa-webfont.woff') format('woff'), url('fonts/spedessa-webfont.ttf') format('truetype')
    font-weight: normal
    font-style: normal
```

and currently invalid for SASS, but much more readable:

```
@font-face
    font-family: 'SPEdessa'
    src: url('fonts/spedessa-webfont.eot')
    src: url('fonts/spedessa-webfont.eot?#iefix') format('embedded-opentype'),
         url('fonts/spedessa-webfont.woff') format('woff'),
         url('fonts/spedessa-webfont.ttf') format('truetype')
    font-weight: normal
    font-style: normal
```

**Comment by olivierlacan** [comment]
I've faced a similar issue yesterday when defining a `@font-face` in Sass.

This feature would have the potential to increase legibility substantially.

Is anyone here familiar enough with how Sass parses lines to have a vague idea how to go about implementing this, or if it's even worth the effort?

**Comment by adamkiss** [comment]
+1

**Comment by mohsen1** [comment]
Can we at least have JavaScript like multi-line with a `\` at end of each line?

``` css
background-image: radial-gradient($color, $color $thickness/2, transparent $thickness/2),\
 radial-gradient($color, $color $thickness/2, transparent $thickness/2)
```

**Comment by firedev** [comment]
Just stumbled onto the super-long line and the first thing I tried was `\` at the end of line, sadly it wasn't working.

**Comment by averyvery** [comment]
:+1:

**Comment by DavidOliver** [comment]
:+1:  It would be very nice for nested lists, too.

```
$list:
    (1,  "value"), (5,  "value"), (23, "value"), (85, "value"), (32, "value"), (11, "value"), (35, "value"), (89, "value")

$list:
    (1,  "value"),
    (5,  "value"),
    (23, "value"),
    (85, "value"),
    (32, "value"),
    (11, "value"),
    (35, "value"),
    (89, "value")
```

**Comment by IanOliver** [comment]
Agreed. Multi-line would be a very welcome addition for mixins and lists.

The whitespace-aware/inferred nature of the indented syntax is the primary reason I chose it over SCSS (and, I assume, it's existence).

As mentioned by others, the readability of box-shadows, text-shadows, @font-faces and lists would benefit hugely from this.

**Comment by lolmaus** [comment]
Selectors too. Sometimes they can be very long.

It's a pity that Sass can't do multiline expressions. The funny thing is that Haml does allow breaking the line after a comma.

**Comment by firedev** [comment]
@lolmaus Here is a sample of my HAML code, and I don't really like it, am I doing it wrong? AFAIK Haml doesn't allow you to simply end a line with comma.

```
  = text_field :inquiry, |
        :customer, |
        placeholder: 'Customer', |
        id: 'suggest_customer', |
        value: (params[:inquiry][:customer] if params.fetch(:inquiry, false)), |
        data: { |
          action: suggest_customer_supervise_or_manage_users_path, |
          email: 'q[name_or_email_cont]' |
        } |
```

**Comment by lolmaus** [comment]
@firedev Does it work when formatted like this?

``` haml
= text_field :inquiry, :customer, placeholder: 'Customer',
                                  id:          'suggest_customer',
                                  value:       (params[:inquiry][:customer] if params.fetch(:inquiry, false)),
                                  data:        { action: suggest_customer_supervise_or_manage_users_path,
                                                 email: 'q[name_or_email_cont]' }
```

A shorter way:

``` haml
= text_field :inquiry,
             :customer,
             placeholder: 'Customer',
             id:          'suggest_customer',
             value:       (params[:inquiry][:customer] if params.fetch(:inquiry, false)),
             data:        { action: suggest_customer_supervise_or_manage_users_path,
                            email: 'q[name_or_email_cont]' }
```

I didn't test it, but i format my Haml code similarly and it works.

TL/DR: You don't need a pipe in the end of a line of Haml code if the line ends with a comma.

**Comment by firedev** [comment]
I'll be damned! It does, what is the trick here? I tried it before and always had to pipe it up.

Guess the problem was caused by me trying to have curly braces on new lines for data:

```
data: {
  ...
}
```

**Comment by pdufour** [comment]
+1

**Comment by mxlje** [comment]
Is there any progress on this?

I’m running into similar issues when trying to use maps in Sass 3.3.5. While when using the scss syntax everything works perfectly like so

``` scss
$susy: (
  columns: 12,
  gutters: 1/3,
  gutter-position: after,
  math: fluid,
  output: float,
  flow: ltr,
  global-box-sizing: border-box
)
```

something like (using `.sass`)

```
$susy:
  columns: 12,
  gutters: 1/3,
  gutter-position: after,
  math: fluid,
  output: float,
  flow: ltr,
  global-box-sizing: border-box
```

doesn’t, because `Nothing may be nested beneath variable declarations.`. While writing it on a single line may work, it’s really hard to read, hits 80 char limits really fast and kind of defeats the purpose of the indented syntax, no?

**Comment by BenjaminHolm** [comment]
I agree with mxlje. The .sass syntax is beautiful but maps really tear things up.

**Comment by Malet** [comment]
Found this thread from http://stackoverflow.com/q/23429137/174034 - I'm having similar problems

**Comment by Arcovion** [comment]
+1

**Comment by starwilly** [comment]
+1

**Comment by nicooprat** [comment]
+1, maybe we should start a bounty for this everlasting issue ?

**Comment by roytomeij** [comment]
Only now find out about this issue. +1

**Comment by skython** [comment]
+1

**Comment by mbixby** [comment]
+1 for bounty

**Comment by vkzawa** [comment]
+1, I just started using the new maps and my markup gets hard to read when defining more than a couple key value pairs.

**Comment by henrahmagix** [comment]
I too would love this feature. I see the `Planned` label is added: is there a timeline for this feature?

**Comment by hauleth** [comment]
:+1:

**Comment by lozandier** [comment]
:thumbsup: +1000 OMG, I thought it was just me. 

Unless it's been discussed before in duplicate issue boxes, has @hcatlin, @nex3, or @hcatlin discussed the biggest caveats of adding this feature that may not be immediately obvious to most?

**Comment by baires** [comment]
:+1:

**Comment by lolmaus** [comment]
@lozandier, no caveats, it's just that the logic behind indented syntax is old and funny nobody wants to get their hands dirty.

**Comment by ghost** [comment]
+1

**Comment by colepeters** [comment]
+100000000000000000000

**Comment by michelegera** [comment]
+1

**Comment by mottosso** [comment]
+1

**Comment by myabc** [comment]
I'd love to get some feedback on #1504.

_for @next3 / the maintainers_
- The PR only improves multi-line support in one specific area: for maps and lists as variables. It won't help with mix-in declarations, multi-line properties, etc. While as a user of the indented syntax, I'd like to see this improvement as soon as possible, I could understand why the maintainers might prefer to hold off and improve support for multi-line across the board. 
- I find the organisation of unit tests in the codebase pretty unclear. Thus I'm not sure if I added tests in the right place.

_for everyone_

How Sass-like should the syntax be?

I worked on the basis of the suggestion from @mxlje (https://github.com/sass/sass/issues/216#issuecomment-41361084). 

``` sass
$susy:
  columns: 12,
  gutters: 1/3,
  gutter-position: after,
  math: fluid
```

Something like this might also be possible though:

``` sass
$susy:
  columns: 12
  gutters: 1/3
  gutter-position: after
  math: fluid
```

In other words, we could make commas optional (or even illegal) - This would give map key-values a very similar syntax to properties. Personally, I'm on the fence about whether this would be a good thing.

n.b. since Sass is pretty liberal about what it allows as a list delimiter, this already works:

``` sass
$stuff:
  "apples"
  "pears"
  "oranges"
```

**Comment by nex3** [comment]
@myabc I'm really glad to see someone looking into this, but I'm not a fan of the way this parsing works. Allowing completely arbitrary expressions but requiring them to be indented beneath variable declarations seems likely to paint us into a corner with respect to backwards compatibility once we do rewrite the indented syntax parser.

> How Sass-like should the syntax be?

My vision for the rewritten parser is that it will support both an indented style and a delimiter-based style, similar to how YAML works. For example, you could write:

``` sass
.class
  .nested {property: value}
  property: value
  property: value
```

I'd like for this to extend to map literals in the way you suggest, although doing so gets a little tricky due to the separation between the statement-level parser and the expression-level parser.

**Comment by hesselbom** [comment]
+1 :+1: :)

**Comment by baires** [comment]
@myabc Great news! This is is a huge step

**Comment by herschel666** [comment]
+1

**Comment by gespinha** [comment]
+1

**Comment by rudeayelo** [comment]
+1

**Comment by Johnius** [comment]
+1

**Comment by foxumon** [comment]
+1

**Comment by chpio** [comment]
+2 :+1: :+1:

**Comment by oreqizer** [comment]
+1337

**Comment by solomonhawk** [comment]
:+1:

**Comment by chpio** [comment]
+42 :+1:

**Comment by greypants** [comment]
:+1:

**Comment by const-int** [comment]
+1

**Comment by HellPat** [comment]
:+1:

**Comment by chriseppstein** [comment]
This is not high on our list for things we're working on, but we will accept a patch that does this.

**Comment by myabc** [comment]
@chriseppstein I did submit a patch (see #1504), but I'll admit that it wasn't a particularly elegant solution. It seems that doing this right hinges on doing a rewrite of the parser for Sass syntax.

I would certainly be up for helping out with this, but it wouldn't be something I'd want to embark on without some guidance from the core team. Embarking on a a rewrite also begs the question, given the interest in [libsass](https://github.com/sass/libsass), as to whether it's worth investing time in a Ruby variant, rather than developing in C++.

**Comment by nex3** [comment]
As with most aspects of Sass design, writing the actual code is going to be a small percentage of the overall effort. The difficult part will be figuring out the syntax and a good algorithm for parsing it; once that algorithm exists, porting it to another language won't be that difficult.

Also, there are currently no plans for Ruby Sass to go anywhere.

**Comment by firedev** [comment]
Sorry, is there any way to write this in SASS?

```
    box-shadow:
        inset 0 1px rgba(white, 0.5),
        inset 0 -5px 10px rgba($blue_shadow, 0.5),
        0 1px 3px rgba(desaturate($blue_highlight, 50%) ,0.5),
        0 0 0 5px rgba(desaturate($blue_shadow, 50%),0.1)
```

Right now I am simply inlining everything:

```
    box-shadow: inset 0 1px rgba(white, 0.5), inset 0 -5px 10px rgba($blue_shadow, 0.5), 0 1px 3px rgba(desaturate($blue_highlight, 50%) ,0.5), 0 0 0 5px rgba(desaturate($blue_shadow, 50%),0.1)
```

**Comment by pylover** [comment]
:+1:  it's good to support `\` just like javascript and python

**Comment by safareli** [comment]
I have been waiting for `\` for about 2 years  :/ and still nothing

**Comment by c9s** [comment]
Maybe it's at the bottom of the list :p

**Comment by safareli** [comment]
and first reason to switch to scss :/

**Comment by roytomeij** [comment]
For what it's worth, and this may be a temporary solution for @safareli too: I mostly want to use this for values of variables (long lists/maps, etc). Currently I keep those in a `_variables.scss` file and `@import` it into my `.sass` files. Keeps the curly brackets and semi-colons to a single file :)

**Comment by baires** [comment]
Im on the same situation @roytomei. I try to keep my maps on an `.scss` file, but with a colons, semicolons, curly brackets etc out of my visual.

Its not perfect, but Im happy with it

``` sass
$theme-colors                                                           : (
  coral: $color-coral                                                   ,
  orange: $color-orange                                                 ,
  yellow: $color-yellow                                                 ,
  green: $color-green                                                   ,
  teal: $color-teal                                                     ,
  blue: $color-blue                                                     ,
  purple: $color-purple                                                 ,
  gray: $color-gray                                                     ,
  black: $color-black
                                                                        );
```

**Comment by kreo** [comment]
+1

**Comment by Johnius** [comment]
No one will fix it because it's pain in the ess to understand the right way to do it?

**Comment by rennhak** [comment]
+1

**Comment by toastal** [comment]
+1

**Comment by MakhBeth** [comment]
+10000000000000
I don't want to drop the indented syntax for this. plz.

**Comment by pylover** [comment]
+1
The indented syntax is awesome! just like haml, yaml and python. So adding multiline expression support  is not a new problem. PLZ

**Comment by adambair** [comment]
+1 - just read the docs and got excited to try out some new stuff -- expected multi-line declarations but no such luck.

**Comment by ragnarkurm** [comment]
+1

**Comment by anthonylebrun** [comment]
+1

**Comment by mlison** [comment]
+1

**Comment by tobynet** [comment]
+1

**Comment by skrzyszewski** [comment]
+1

**Comment by scottbaggett** [comment]
+1

**Comment by gregkohn** [comment]
+1

**Comment by saeedtavoosi** [comment]
+1

**Comment by harrislapiroff** [comment]
+1. I've started a bounty on this issue. I put $15 on it. I hope others will do the same! https://www.bountysource.com/issues/395405-multiline-expressions-in-indented-syntax

**Comment by scottbaggett** [comment]
I've added $30 to the bounty bringing it up to $45. Let's get this done!

**Comment by lolmaus** [comment]
Another $5 from me, $50 total.

**Comment by chriseppstein** [comment]
It would be really great if we could make the indented syntax parser a simple pre-parsing step before handing it off to the SCSS parser.

**Comment by HamptonMakes** [comment]
Kind of like how libsass does it?

—

**Comment by chriseppstein** [comment]
> Kind of like how libsass does it?

@hcatlin I've not read the details of that implementation yet, so I can't say for sure. But, at a high level, yes. This was the idea I floated for supporting the indented syntax during when we created SCSS, but @nex3 had some reservations about the implementation I suggested. I have since had some ideas that would not have the downsides that my original implementation approach would have had.

**Comment by lukewatts** [comment]
Does stylus do this? I know Jade implemented this for their javascript variable declarations.

So this works now (in Jade):

``` javascript
-
    var array = [
        'item1',
        'item2',
        'item3'
    ]
```

Whereas it never did before. Perhaps that will spark some ideas.

That said, I'm totally in favor of a "Python-esque" solution

**Comment by nscarcella** [comment]
+1 I want this so much!

**Comment by typical000** [comment]
+1 Because in SASS it's so hard to write something like this:

``` sass
$colorSchemes: (
  'black': (
    'outerBorderColor': #141414,
    'textColor': #bbb,
    'bgBlockFirst': #101010,
    'bgBlockSecond': #0c0c0c,
    'linesStroke': #202020,
    'dotsStroke': #202020,
    'dotsFill': #303030,
    'arrowFill': #aaa
  ),
  'sand': (
    'outerBorderColor': #786e52,
    'textColor': #e5e5e5,
    'bgBlockFirst': #908667,
    'bgBlockSecond': #897e5d,
    'linesStroke': #9a9074,
    'dotsStroke': #9a9074,
    'dotsFill': #a59e86,
    'arrowFill': #fff
  )
);
```

**Comment by kirahowe** [comment]
Any word on this becoming a reality? I'll add yet another +1.

**Comment by scottbaggett** [comment]
Add more $$ to the bounty. :)

**Comment by ellioseven** [comment]
It's stuff like this that makes me move to Stylus.

**Comment by harrislapiroff** [comment]
I added more :moneybag: to the bounty! I encourage anyone else who wants this to do the same. Let's make it worth someone's while, since it sounds like a tricky thing to implement. :) https://www.bountysource.com/issues/395405-multiline-expressions-in-indented-syntax

**Comment by chriseppstein** [comment]
> It's stuff like this that makes me move to Stylus. This has been an issue for 4/5 years!

@ellioseven Hi, I sent you a tweet. I'd like to chat with you if you have a few minutes.

**Comment by ranska** [comment]
+1

**Comment by igoradamenko** [comment]
+1. it will be great for instance for box-shadows:

```
box-shadow: 0 5px 5px -5px #fff,
            5px 0 5px -5px #fff
```

**Comment by chriseppstein** [comment]
Depends on #1942

**Comment by davidgoli** [comment]
👍

**Comment by simonvomeyser** [comment]
Another :thumbsup: ... big fan of the indented syntax, but maps are not working

**Comment by maciej-ka** [comment]
Hi @simonvomeyser. Could you write an example?

**Comment by simonvomeyser** [comment]
Hey @lokson, came here from #1088 "Multiline declaration of maps in SASS" - Something like this is not working with the sass-syntax.

```
$layout: (
  columns: 12,
  gutters: 20px
);
```

these maps have to be in one line to work like so

```
$layout: (columns: 12, gutters: 20px, ...)
```

This seems to be known and I just wanted to give my +1 here. Thanks for asking tough! :)

**Comment by maciej-ka** [comment]
@simonvomeyser 

I believe multiline operator is going to be a part of upcoming Sass v3.5 (https://github.com/sass/sass/pull/2094). After some discussion on the syntax the result is:

``` sass
$layout: ( \
  columns: 12, \
  gutters: 20px \
);
```

**Comment by lukewatts** [comment]
That's better than nothing. Like I mentioned before Jade (which is an indented html templating language) uses a multiline operator BEFORE the chunk of code in question. Maybe something like this would involve less typing:

```
\\
    $layout: (
        columns: 12,
        gutters: 20px    
    )
```

That way anything nested after the `\\` would be set to be multiline...

**Comment by igoradamenko** [comment]
@lukewatts, what if the chuck of code is at the beginning of a file?

**Comment by maciej-ka** [comment]
@lukewatts there was ready implementation for something similar (check https://github.com/sass/sass/pull/2094). It would use indentation to join lines, like:

``` sass
$layout: ( \
  columns: 12,
  gutters: 20px
  )
```

But this approach was voted as untypical and misleading, so it was changed into:

``` sass
$layout: ( \
  columns: 12, \
  gutters: 20px \
);
```

**Comment by harrislapiroff** [comment]
Was there consideration of also doing something like Python's implied line continuation which allows any line breaks within parens without the need for a backslash?

I'm definitely happy with the solution as implemented now, but implied line continuation seems like it would be intuitive in most circumstances.

**Comment by lukewatts** [comment]
@igoradamenko It wouldn't make a difference. Only items "nested" after the `\\` (or whatever special character) would be interpreted as multiline. For example:

```
\\
    $susy: (
        gutters: 1/4,
        columns: 12
    )

$not-a-multiline-map: (some: value, go: here)

.not-multiline-because-not-indented-after-special-char
    display: block
    padding: 1em
```

Same as comments. If you have a `//` at the beginning and you nest something under that it doesn't mean you're whole file is commented out unless everything is indented one level after the comment.

**Comment by lolmaus** [comment]
This is ridiculous. Good indentation-driven languages like [Jade](http://jade-lang.com/), CoffeeScript, Elm do not require any hints from the user.

**Comment by lukewatts** [comment]
I agree with @lolmaus Ideally less is more. What if the brackets weren't even required? Why not this:

```
$map:
    columns: 12
    gutters: 1/4
```

Isn't that what cson does for json? I mean it's indented. Make indents matter, not symbols

**Comment by ragnarkurm** [comment]
Why I prefer sass over scss is less typing.
Less markup means cleaner code, easier to grasp by eye, less mistakes.

**Comment by ghost** [comment]
+1 sass doesn't seem awesome anymore without this feature

**Comment by magnetospheric** [comment]
+1 lack of this feature is making my stylesheet declarations absurdly long

**Comment by delijah** [comment]
+1

**Comment by iErik** [comment]
Is there any workaround for this problem for now?

**Comment by neo-eureka** [comment]
That's really funny: the issue is open for **more than 5 years (!)** with a hundred of users asking to implement a (simple) feature, and it looks like nobody cares. Is SASS still actively maintained? This feature is really not so hard to implement compared to all the rest of SASS features implemented. What is the sense of open source software then? We have a million libraries/frameworks out there (cool!), but when you try to use them they all either do not work, or do not work as you expected, or have qualities which render them totally useless... It's sad and silly...

**Comment by mlison** [comment]
@neo-eureka one of the main pillars of open source software is that anyone can contribute to the project. In my opinion this "issue" is largely cosmetic, a "nice-to-have", and doesn't prevent you from actually using the tool.

Perhaps next time instead of writing an elaborate rant like that you could consider making a pull request?

**Comment by pylover** [comment]
@neo-eureka , It's very hard for anyone, who comes from outside of a context, to dive into, and fix such problems. I think the authors should responsible to their communities. 

A lot of users are waiting for a simple feature. It seems after 5 years, There is no body to fix it. this approves my reason. because we can not do that within a simple trick. It's take a lot of time to i understand the project structure and etc ...

**Comment by pylover** [comment]
@neo-eureka , Currently you are using a lot of software made by open source developers, and they are very responsible to their mistakes and feature requests. That's why the open-source world is working for all of us

**Comment by mlison** [comment]
@pylover your argument depends on an assumption that "if somebody is the only one who can do it, he is obligated to do it", and that is, simply, not true.

**Comment by pylover** [comment]
@mlison, Other users will judge about our ideas.

**Comment by mlison** [comment]
@pylover Now don't get me wrong, I too prefer projects without issues. The responsibility that you are referring to, though, is only a personal responsibility. It is a choice of whether and how you are going to deal with such problems. The core of your idea is basically "if you can't do it properly, don't do it at all" - with that, we would all be very surprised how much open source software wouldn't exist. Publishing software on Github is not a contract between the developer and rest of the world.

**Comment by myabc** [comment]
@neo-eureka @pylover it is not a "simple" feature.

I came up with a hacky implementation a couple of years ago https://github.com/sass/sass/issues/216#issuecomment-61413816. Unfortunately, in addition to being a hack, I didn't get much feedback about the syntax I proposed. There are many who have +1'd this issue, but it seems there are relatively few in the community who are willing to dig in to the specification and implementation details.

**Comment by maciej-ka** [comment]
@neo-eureka Ruby Sass is not maintained anymore and multiline feature has been closed without merge: https://github.com/sass/sass/pull/2094

**Comment by iErik** [comment]
@lokson Well, If I'm not mistaken, there's still node-sass around, right? Anyhow, I don't hold any hopes for the Indented Syntax anymore, for years it seems like they've been trying to kill it anyways, it's just not a thing anymore, despite being much more cleaner and elegant than raw CSS or even SCSS. Personally, I do prefer the Indented Syntax over the Non-Indented one (scss), but maybe scss is just much more simpler to maintain and to learn (even though is not that difficult to learn the Indented Syntax, honestly).

**Comment by toastal** [comment]
There's been attempts to kill it? Seems [loud and clear](http://thesassway.com/news/sass-is-here-to-stay) that it's not going anywhere and all of the examples on the site still use both. Tools targeted specifically at Sass work for both (or you should raise a bug issue). All new features are expected to work in both syntaxes, and they do -- they just have to be on one line. I'd like this lack of multi-line support fixed too, but I feel like there's a lot of overreaction.

**Comment by audunolsen** [comment]
+1

**Comment by pylover** [comment]
I love indented syntax and can not imagine the sass with `}{;`

**Comment by CyberCookie** [comment]
+1

**Comment by Timofey81** [comment]
+1

**Comment by wouter-muller** [comment]
+1

**Comment by tyburrowbridge** [comment]
+1

**Comment by Gotterbild** [comment]
+1

**Comment by ghost** [comment]
+1

**Comment by IvRRimum** [comment]
+1

**Comment by iErik** [comment]
+1

**Comment by philpoore** [comment]
+1

**Comment by some1else** [comment]
Man..

**Comment by factreset** [comment]
+Infinity

**Comment by joallard** [comment]
Dear "+1" spammers: can we please take actual steps towards resolving this instead of just adding to the noise with yet another empty comment? Looking at you @factreset @some1else @philpoore @iErik @IvRRimum ...

It is utterly useless at solving the issue and just triggers a notification for the 91 people in here for absolutely no reason. Looks like the actual way forward is making improvements on dart-sass and the spec in collaboration with the maintainers.

**Comment by wrwrwr** [comment]
Please excuse the noise. For those that happen to use webpack and do not mind an imperfect, temporary solution: [webpack-multiline-sass](https://github.com/wrwrwr/webpack-multiline-sass).

**Comment by Lukortech** [comment]
Hi! 
I am expiriencing a problem with scss syntax being highlited as 'wrong'.
That is: `grid-template-columns: repeat(6, 1fr);`
being turned into: `-ms-grid-columns: (1fr)[6];
      grid-template-columns: repeat(6, 1fr);`

The output from VS code (`] expected`, `semi-colon expected`, `{ expected`, `at-rule or selector expected`) says not much for me, anyone care to help?

**Comment by henrahmagix** [comment]
Hi @Lukortech, I don’t think this is the right place for your question, but I’d be happy to help you elsewhere :) if you could send me a tweet with some details about the plugins you’re using in VS Code, I’m sure we can figure this out: my twitter handle is the same as my github one :)

**Comment by Lukortech** [comment]
Hi, I am sorry for not replying. 
Seems like I missed your message. 
I managed to get through it, if I'll find anything more, will let you know via. Twitter ;) 
Have a good one!

**Comment by LilyReile** [comment]
This would be nice. Our linter is not going to like this monster:
```
transition: all $base-transition-timing $base-easing $base-transition-timing, opacity $base-transition-timing $base-easing 0s;
```

**Comment by agronick** [comment]
Its been 8 years!! When is this going to happen?

**Comment by neo-eureka** [comment]
> Its been 8 years!! When is this going to happen?

Sorry to say that, but maybe the answer is NEVER. Welcome to FOSS.
Feel free to fork it, own it and implement any features you need yourself.

**Comment by shreve** [comment]
> Yes, this is still a language feature we'd like to add. If someone wants to put together pull requests to add specs for it to [sass-spec](https://github.com/sass/sass-spec) and an implementation to Dart Sass, I'd accept them. The language feature should still be tracked in [sass/sass#216](https://github.com/sass/sass/issues/216), though.
> -- [nex3](https://github.com/sass/dart-sass/issues/280#issuecomment-378773350)

If you want this feature to exist:

1. Head over to [sass-spec](https://github.com/sass/sass-spec) and make a PR with some specs indicating how this feature should work.
2. Either write the feature yourself in [Dart sass](https://github.com/sass/dart-sass) or find someone who can.

That's it. Dart sass is now the [primary implementation](https://sass-lang.com/dart-sass) of sass. Once it's there, it will make it's way to the others.

**Comment by toastal** [comment]
I think it should also be mentioned as an alternative to use PostCSS + [SugarSS](https://github.com/postcss/sugarss). You can get your multiline breaks and without the proposed `\`. After following this bug for years and no one taking the torch, SugarSS is what I've switched to personally and at work. As you probably already know, there's a healthy amount of tools in the PostCSS toolchain to do most of the task you want like nesting, etc. -- and you're probably already using it already if you use Autoprefixer -- without the Sass compiler dependency. Sure, you don't get access to Bootstrap or some of these other big frameworks, but that's probably okay for many.

---

Perspective change from 2019 to 2022: indented syntax isn't worth it (even if it's easier to read/write). Sass isn't worth it either. Unless we are library writers--who should be targeting CSS--we can achieve 95% of what we need in modern CSS's features (like variables, etc.), better naming schemes, and an even smaller toolchain, like LightningCSS. When it comes to keeping code maintainable and accessible to developers of all backgrounds, sticking with the base syntax should be preferred unless there are big lacking features and considering how much complexity (not to mention broken "features", like all of Sass's color functions, `image-set()`, etc.) this tool brings, it should be avoided. CSS should be simple so that juniors and back-end developers can hop in without having to know much.

---

2025 perspective: library authors still deserve nice tools too & with dart-sass being decoupled from the mess that is npm dependency hell, I would rather avoid a complex PostCSS toolchain in favor of Sass at this point—just not over using features that aren’t needed. This perspective comes from doing a lot of Nix where JavaScript packaging really rears its ugly head & I try to avoid it.

**Comment by lolmaus** [comment]
A humble reminder that there's a bounty on this feature reqeust: https://www.bountysource.com/issues/395405-multiline-expressions-in-indented-syntax

If you're interested in seeing multiline indented syntax implemented, please contribute to the bounty.

**Comment by agronick** [comment]
Correct me if I'm wrong, but wouldn't this just be a matter of running a regular expression on the file to replace any backslash followed by a linebreak with an empty string? Seems like it would be easy with not a lot of edge cases.

**Comment by lolmaus** [comment]
@agronick That could be a legit workaround, but we deserve first-class multiline support. 

There has never been any multiline issue with CoffeeScript, Haml, Python. Sass is really unfortunate here.

**Comment by j-holub** [comment]
It's beyond me how something like this is not implemented. I'm a long time SCSS, first time SASS user and I really loved the clean synthax. Now I got to the part where I use `grid-template-area` to define my image galleries that have a lot of rows and the SASS way of writing everything into a single line, makes it completely unreadable.

**Comment by joallard** [comment]
Folks, we get it. Everyone does. It is a legitimate need.

Please, instead of reminding everyone how it's unfair that it doesn't exist yet, take a look at [comments above](https://github.com/sass/sass/issues/216#issuecomment-504441260) and invest your energy towards resolving this problem, one step at a time.

**Comment by CyberCookie** [comment]
@joallard
as @agronick have mentioned it could be solved easily just having preprocessed every .sass file with regular expression: you should add special symbol to mark line as a new line and then process it with regexp to replace occurrences with spaces
sassFile.replace(/ SPECIAL_SYMBOL \n/gm, ' ')

Maybe it will save you another 8 years

**Comment by winston-trey-wilkinson** [comment]
Can I genuinely ask why this, after almost 10 years, has not been solved? Is it due to dart being a difficult language, people just too lazy to fix it, or worry of breaking the rest of the code? I'm considering fixing this absurd issue however I will apparently have to learn dart first. Any reason I shouldn't?

**Comment by lolmaus** [comment]
@winston-trey-wilkinson It's because Sass is an open-source project, not backed by any large company or corporation.

There are few individuals capable of implementing this. The amount of time they devote to OSS is limited, and they are free to set their own priorities. Apparently, none of them is personally interested in this feature.

I'm 100% with you here, but *unfortunately* whining does not help to persuade those people shift their priorities towards this feature.

There's a croudfunding effort here started by @harrislapiroff five years ago: https://www.bountysource.com/issues/395405-multiline-expressions-in-indented-syntax

If every commenter on this thread dropped a dollar, it would be worth $1000 already. If everyone contributed $5 (that's one coffee or one beer), that would amount for $5000, which would definitely be enough to get a capable developer interested in implementing this.

But so far only five people have contributed. Can we really blame Sass maintainers for lack of support, given we don't provide any support either?

**Comment by scottbaggett** [comment]
Agreed.  I put in $30 way back in 2015 so feel free to step up everyone!

**Comment by winston-trey-wilkinson** [comment]
I have decided to take a stab at solving this issue. My goal is a similar pattern to [sugarss](https://github.com/postcss/sugarss#multiline). I do not know any dart (in fact I didn't know it existed until two days ago) and my javascript knowledge is limited but regardless I think this issue is worth solving. This issue is seemingly simple however from other peoples views I think it will be anything but simple. I hope to have the issue at least partially solved by the end of the month but who knows. 
Wish me luck!

**Comment by algomaster99** [comment]
@winston-trey-wilkinson How is the progress?

**Comment by ScamCast** [comment]
Here's a few work-arounds I came up with when trying to figure out how to get my grid areas on multiple lines. Hope a real solution is implemented sometime. I really don't want to leave sass :(


**sass**

    $my-grid:'\
      "col1-row1 col2-row2 col3-row3"\
      "col1-row2 col2-row2 col3-row2"\
      "col1-row3 col2-row3 col3-row3"'
    
    .container
        display: grid
        grid-template-areas: unquote($my-grid)

**css**

    .container {
        display: grid;
        grid-template-areas:   "col1-row1 col2-row2 col3-row3"  "col1-row2 col2-row2 col3-row2"  "col1-row3 col2-row3 col3-row3";
    }
______________________________

**sass**

    .container
        display: grid
        grid-template-areas: unquote('"col1-row1 col2-row2 col3-row3" \
                                      "col1-row2 col2-row2 col3-row2" \
                                      "col1-row3 col2-row3 col3-row3"')


**css**

    .container {
        display: grid;
        grid-template-areas: "col1-row1 col2-row2 col3-row3" 								  "col1-row2 col2-row2 col3-row2" 								  "col1-row3 col2-row3 col3-row3";
    }

**Comment by ranska** [comment]
> @pengwynn I like that approach. It's nicer than the way I was thinking, which was adding a \ to the end of line like in Python.

in slim when you embed ruby code both are possible.
for example.

```slim

= render partial: "super_sass_style_guide", 
  locals: {\
    semico: nil,
  }
```

**Comment by some1else** [comment]
Seems like this would be a straightforward solution, in line with the way some shells and other programming languages wrap long strings :

```
body
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', \
    'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', \
    'Helvetica Neue', sans-serif
```

I've opened [an issue](https://github.com/sass/sass-spec/issues/1557) in the sass-spec repo and added $20 to the [bounty](https://www.bountysource.com/issues/395405-multiline-expressions-in-indented-syntax).

**Comment by lolmaus** [comment]
It's a reasonable workaround and I'm looking forward to it becoming available. :+1: 

But it's not a solution. :grimacing:

**Comment by t-wilkinson** [comment]
Although I enjoyed Sass, I have been looking for alternate solutions. Here are my preferences for those interested: 
1. React. It is encouraged to componatize elements in the React syntax, instead of leaning on css. I much rather prefer its design philosophy and the seperation of concerns (components, not filetypes).
2. Tailwindcss. It is a wonderful solution on its own, but even better with React. It's essentially css in classes using a 'functional'  paradigm. 
3. Expanding horizons. Sass was one of the first css preprocessors I used and I was looking to find new solutions.

With these two elements, I find my code easier to manage and reuse. I only use ~200 lines of css per project. 

This of course does not excuse me for leading people on for so long thinking I was creating a solution. In all honesty, I forgot about my promises, I apologize.

**Comment by hstaniszewski** [comment]
After nine years, SASS still does not support it

**Comment by factreset** [comment]
Just use SCSS guys. With code completion by means of Emmet it doesn't take any longer to type, and the presence of braces and semicolons means:

1. This multiline issue doesn't exist.
2. Nested code is more readable. 
2. CSS can be copy-pasted without the need to edit it to account for indented syntax idiosyncrasies.  

I think for these reasons most developers prefer SCSS anyway, as when working in teams a full-stack dev can open and read an SCSS file without the need to mentally adjust to the indented syntax. Imagine this is the real reason nobody with the necessary level of skill has stepped in to solve the multiline problem.

**Comment by wouter-muller** [comment]
I love the indented Sass syntax. I am willing to accept the multiline issue as an acceptable flaw in an across the board amazing and imo better syntax. If people are interested to read more about SASS vs. SCSS, check this blog i've written: https://woutermuller.net/why-i-prefer-the-sass-indented-syntax-over-the-scss-syntax (sorry if this is off-topic but i felt someone should stand up for Sass, a lot of people are using it and loving it).

Also, @factreset I'd like to encourage you to think about the use of the word "guys". It is in fact sexist (not everyone here is a guy), and using it could lead to women (that we in tech desperately need more of imo) feeling left out. A more inclusive word would be "everyone". 

Back on the topic, my workaround has been like this:

As i like to keep my SASS partials as short as possible, for the rare cases where i need to really use a multiline statement, i just simply use an .scss partial. So >99% of my partials are .sass, <1% is .scss. It compiles out of the box without any extra setup.

**Comment by wouter-muller** [comment]
Back on the topic, my workaround has been like this:

As i like to keep my SASS partials as short as possible, for the rare cases where i need to really use a multiline statement, i just simply use an .scss partial. So >99% of my partials are .sass, <1% is .scss. It compiles out of the box without any extra setup.

**Comment by alromh87** [comment]
Feels like it's time to get this solved :wink: 

You can try it from https://github.com/alromh87/dart-sass/tree/multiline

1. Install dart
2. Clone repo git clone https://github.com/alromh87/dart-sass/tree/multiline
3. Go into dir dart-sass
4. Install dependencies
`pub get`
5. Run 
` dart dart bin/sass.dart testN.sass`

You can use this file as reference, I tested the examples I found here
https://github.com/alromh87/sass-spec/blob/master/spec/multiline_sass/input.sass

**Comment by ngdangtu-vn** [comment]
@alromh87 
Is it possible for you use the comma as separate line instead of `/`?

**Comment by alromh87** [comment]
@ngdangtu-vn while possible it would be more complicated and I wasn't aware that all this [has to go over the language proposal process](https://github.com/sass/dart-sass/pull/1098#issuecomment-703863444), so for now I will wait to see what comes out from this process.

I didn't expect this fix to be that complex 😅

**Comment by ngdangtu-vn** [comment]
@alromh87 Well, it can be help. I think if your PR got approved, it involves to a lot developers. And one it got publish, there will be no way to turn back. So let's way for it got approved. Anyway, thank you for your PR. I do hope it will pass soon 'cause I truly need it.

**Comment by mangelozzi** [comment]
> @alromh87
> Is it possible for you use the comma as separate line instead of `/`?

A comma would be confusing, list and maps are separeted by commas.
One of the most useful reason for multiline statements is for list and maps, which would then require two commas.
Is the comma a comman between items, or to join the line?? Hence why many language use a `\`

**Comment by ngdangtu-vn** [comment]
@mangelozzi 
> Is the comma a comman between items, or to join the line??

Can it do both? I don't understand why should we concern about it.

**Comment by alromh87** [comment]
It would be possible to avoid using any extra character (aka: detecting the coma) but since the parser works on a char level it  would need to keep state when it should treat newline as single line.

For that it would be needed to specify where this behaviour should thake place. Would someone be willing to analyze and integrate one single list? Again I feel it make it more complex and probably will delay more the integration.

I sent proposal as implemented, lets wait an see the result, hopefully won't thake that long since it is allready a very long awaited feature (2011).

In the meantime you can follow the procedure detailed [before](https://github.com/sass/sass/issues/216#issuecomment-699600890) to use it if you strongly need it now. Let me know if you have questions.

Also if someone would be willing contribute to the bounty to support development, I would appreciate it.

One more thing: once upon a time [there was an implementation of this functionality for ruby sass ](https://github.com/sass/sass/pull/2094) that never landed due to unfortunate circumstances, this implementation also used backslash "\\", so should not be too hard to syncronize ruby version as well. I hope unfortunate circumstances dont arrise again, and this issue can be finally closed.

Good vibes :call_me_hand:

**Comment by alromh87** [comment]
Hello, do any of you have any idea how this should be written?

https://github.com/sass/sass/pull/2921#pullrequestreview-507778242

**Comment by LandonSchropp** [comment]
Just to add another use case to this, this would be very helpful when using `grid-template-areas`:

```
grid-template-areas: 
  "primary-navigation primary-navigation" 
  "secondary-navigation sidebar" 
  "main sidebar" 
  "content-controls sidebar"
```

While this wouldn't be my first choice of syntax, as noted by other people, many other languages support this by using backslashes:

```
grid-template-areas: \
  "primary-navigation primary-navigation" \
  "secondary-navigation sidebar" \
  "main sidebar" \
  "content-controls sidebar"
```

**Comment by avesus** [comment]
What are the objective benefits of using SASS vs SCSS again, despite of this awkward yaml-style tabulation?

+1

**Comment by wouter-muller** [comment]
- it is cleaner and slightly easier to read and easier/faster to type, cause there is no need for the {, } and ; symbols
- Easier/shorter way of calling mixins, +mixin instead of @import mixin

@avesus

**Comment by Jorge-Gonzalez** [comment]
It is more consice, and compact, easier to read, IMO.

**Comment by EzraWolf** [comment]
Not trying to be rude, but is there a reason why this still hasn't arrived yet..? It has been so many years since this was first opened

**Comment by HamptonMakes** [comment]
@EzraWolf "Intended Sass" isn't actively being worked on or expanded– it's hardly known at all outside of the Ruby community who got used to it with the original design of Sass wayyy back over 16 years ago when that was the original style.

Over a decade ago we decided to pivot hard to using a 'css-compatible' style of Sass, and have made that the primary syntax ever since and has contributed to the expansion of the language.

In fact, I think it's kinda wild that Natalie has still kept the indented syntax functional after all these years of new features.

I'm not sure if @nex3 has officially commented on this ticket– so many comments to wade through– but I'd guess that she isn't really _against_ it, but likely it's just not even anywhere near a priority with the usage that indented-syntax gets.

Honestly, it causes so much confusion and double-work (this ticket being an example) that I'm not sure how much longer it's even a good idea to keep it around!

**Comment by wouter-muller** [comment]
@HamptonMakes We love the indented syntax, I use it for all my projects (non Ruby or RoR projects but Vite apps and pure Vue SPA's), so please keep it around, we don't care about this multiline issue :)

**Comment by nex3** [comment]
To be clear: we have no plans to get rid of the indented syntax. I would definitely like to add multiline expressions—note that this issue is marked as "planned"—but it'll take a substantial design effort to ensure we get this right and the Sass team is stretched very thin.

**Comment by HamptonMakes** [comment]
What she said then! ^^

**Comment by rossbulat** [comment]
Sass syntax works very well with Tailwind CSS, this is giving developers a fresh perspective of whether to support indented syntax. [Tailwind sass syntax](https://marketplace.visualstudio.com/items?itemName=macieklad.tailwind-sass-syntax) (syntax highlighter only) for VS Code has had 37,086 installs at time of writing, and while not a huge amount, there is an obvious desire for the combo.

**Comment by j-tap** [comment]
Thank you, we look forward to this opportunity

**Comment by geoidesic** [comment]
Wow, 10 years later and we still have this issue!?
It's pretty major for any code that uses gradients.

@HamptonMakes writing semi-colons and braces is what I call `double-work`. Normalise indented syntax it's just objectively better. Python knows this. Pug knows this. Sass should know this. Why is it objectively better? Because the most important virtual estate in coding is vertical, as a result, braces reduce the amount of code you can see on screen, which slows you down as a developer. Not just that but it's far, far easier to restructure nesting in indented code than in code with braces. As a developer building stuff for user-land, the primary metric of success is speed to market. So this stuff (indent syntax) matters!

**Comment by nabato** [comment]
Sass syntax is absolutely wonderful. Are there any plans on implementing this feature?

**Comment by Way** [comment]
Is there any chance or even plan that the said "planned" feature of multi-line support will come at some point?

**Comment by nex3** [comment]
A [proposal for this](https://github.com/sass/sass/blob/main/proposal/indented-syntax-improvements.md) is out for review, along with [an explainer blog post](https://sass-lang.com/blog/rfc-indented-syntax/). Since this is a substantial but non-breaking change, I'll give it a month for public comment before we mark it as accepted.

**Comment by jamesnw** [comment]
In #3991, tooling authors are raising concerns about the implementation cost of the proposed [SCSS-in-Sass](https://sass-lang.com/blog/rfc-indented-syntax/#scss-in-sass) portion. For authors using the indented syntax, how helpful would the ability to have inline SCSS statements in your `.sass` files.

Some key benefits would be terser format for short styles-

```
.foo {color: red}
```

and a more familiar syntax for multiline statements. Much of what can be accomplished with SCSS-in-Sass can also be done by wrapping statements in parentheses. For example, this would be possible under the proposal without SCSS-in-Sass-

```
.foo
  box-shadow: (
    inset 0 1px rgba(white, 0.5),
    0 0 0 5px rgba(blue, 0.1)
  )
  grid-template-areas: (
    "primary-navigation primary-navigation" 
    "content sidebar"
  )
```

SCSS-in-Sass would allow the following inside of a `.sass` file-

```
.foo {
  box-shadow: 
    inset 0 1px rgba(white, 0.5),
    0 0 0 5px rgba(blue, 0.1);
  grid-template-areas: 
    "primary-navigation primary-navigation" 
    "content sidebar";
}
```

Thoughts? Do you have situations where you'd want to reach for SCSS-in-Sass?

Feel free to comment, or vote on this comment, with 👍 indicating you would find it useful, or 😕 if you don't think you would use it.

**Comment by voltaek** [comment]
It seems to me that one solution to the original issue of multiline expressions in the indented syntax is plenty, especially when it is a simpler solution than SCSS-in-Sass and doesn't introduce tooling complexities and/or the potential for user confusion that mixing syntaxes does.

**Comment by geoidesic** [comment]
No one wants scss in sass indented mode. The whole point of of sass indented is to avoid the extra punctuation!

**Comment by KBeDevel** [comment]
At least for me, the point of using indented syntax is to have a lighter and easy-to-read style files. Not exclusively, SCSS-in-Sass make sense for styles migration purposes.

**Comment by nex3** [comment]
It's been two month and it seems like the non-SCSS-syntax parts of the proposal are uncontroversial, so I'm going to mark it as approved.

**Comment by nex3** [comment]
This has now landed and been released in 1.84.0!

**Comment by lolmaus** [comment]
Amazing!

This feature request is 13 years old! It's older than my son! 🙈 

Back in the day, I've been a huge proponent of indented syntax, with CoffeeScript and all. 😅 I honestly did not believe I'll ever see this feature landing in Sass.

Thanks, @nex3! 🙇

**Comment by Lukortech** [comment]
> Amazing!
> 
> This feature request is 13 years old! It's older than my son! 🙈 
> 
> Back in the day, I've been a huge proponent of indented syntax, with CoffeeScript and all. 😅 I honestly did not believe I'll ever see this feature landing in Sass.
> 
> Thanks, @nex3! 🙇 

I've got two kids now, when I initially wanted that request they were not even in the picture yet. Time flies. 
Good thing to see such issues resolved... Eventually :D

**Comment by scottbaggett** [comment]
I see my $30 bounty was well spent.  Congrats everyone! :)

**Comment by BarnabasSzabolcs** [comment]
So, can I now write box-shadow's, too, multiline?
```
box-shadow: 
  0 4px 6px rgba(0, 0, 0, 0.1), // Soft main shadow
  0 1px 3px rgba(0, 0, 0, 0.05)  // Subtle inner shadow
```
or something like this maybe with parentheses?

**Comment by jamesnw** [comment]
> So, can I now write box-shadow's, too, multiline?
> 
> ```
> box-shadow: 
>   0 4px 6px rgba(0, 0, 0, 0.1), // Soft main shadow
>   0 1px 3px rgba(0, 0, 0, 0.05)  // Subtle inner shadow
> ```
> 
> or something like this maybe with parentheses?

You can- you'll just need to wrap the box shadow in parentheses, like this-

```sass
a
  box-shadow: (
    0 4px 6px rgba(0, 0, 0, 0.1), // Soft main shadow
    0 1px 3px rgba(0, 0, 0, 0.05)  // Subtle inner shadow
  )
```

You can explore more on the [playground](https://sass-lang.com/playground/#eJwzMEzkUlBIyq/QLc5ITMkvt1LQAPIVFAwUTAoqFMyAuCg9KVHDQEcBivQMNXUU9PUVgvPTShRyEzPzFCAaoboMgTqMsegyMNVUAGsrTSrJSVXIzMtLLULo1AQARZAjlA==).
